ql/src/java/org/apache/hadoop/hive/ql/metadata/client/LocalCachingMetaStoreClientProxy.java

```diff
@@ -0,0 +1,465 @@
+import org.apache.hadoop.hive.metastore.api.AggrStats;
+import org.apache.hadoop.hive.metastore.api.ColumnStatisticsObj;
+import org.apache.hadoop.hive.metastore.api.GetPartitionsByNamesRequest;
+import org.apache.hadoop.hive.metastore.api.GetPartitionsByNamesResult;
+import org.apache.hadoop.hive.metastore.api.GetTableRequest;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.apache.hadoop.hive.metastore.api.Partition;
+import org.apache.hadoop.hive.metastore.api.PartitionSpec;
+import org.apache.hadoop.hive.metastore.api.PartitionsByExprRequest;
+import org.apache.hadoop.hive.metastore.api.PartitionsStatsRequest;
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.client.HiveMetaStoreClientUtils;
+import org.apache.hadoop.hive.metastore.client.BaseMetaStoreClientProxy;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+import org.apache.hadoop.hive.ql.metadata.Hive;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.CacheKey;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.CacheWrapper;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.KeyType;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.PartitionNamesWrapper;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.PartitionSpecsWrapper;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.PartitionsWrapper;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.Supplier;
+import org.apache.hadoop.hive.ql.metadata.client.MetaStoreClientCacheUtils.TableWatermark;
+import org.apache.hadoop.hive.ql.session.SessionState;
+import org.apache.hadoop.hive.ql.util.IncrementalObjectSizeEstimator.ObjectEstimator;
+import org.apache.hadoop.hive.ql.util.IncrementalObjectSizeEstimator;
+import org.apache.thrift.TException;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.hadoop.hive.metastore.utils.MetaStoreUtils.parseDbName;
+
+/**
+ * This class introduces a caching layer in HS2 for metadata for some selected query APIs. It extends
+ * HiveMetaStoreClient, and overrides some of its methods to add this feature.
+ * Its design is simple, relying on snapshot information being queried to cache and invalidate the metadata.
+ * It helps to reduce the time spent in compilation by using HS2 memory more effectively, and it allows to
+ * improve HMS throughput for multi-tenant workloads by reducing the number of calls it needs to serve.
+ */
+public class LocalCachingMetaStoreClientProxy extends BaseMetaStoreClientProxy
```

Comment:
why we keep adding `implements IMetaStoreClient` everywhere where we have `extends BaseMetaStoreClientProxy`? 
BaseMetaStoreClientProxy extends NormalizedMetaStoreClient that implements IMetaStoreClient